File: code_snippets/go/usecases/workflows/workflows.go
package main

import (
	"context"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log"
)

type User struct {
	Name  string
	Email string
}

// <start_here>
type SignupWorkflow struct{}

// <mark_1>
func (SignupWorkflow) Run(ctx restate.WorkflowContext, user User) (bool, error) {
	// </mark_1>
	// workflow ID = user ID; workflow runs once per user
	userId := restate.Key(ctx)

	// <mark_2>
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return restate.Void{}, CreateUserEntry(user)
		// </mark_2>
	}); err != nil {
		return false, err
	}

	// <mark_2>
	secret := restate.Rand(ctx).UUID().String()
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return restate.Void{}, SendEmailWithLink(userId, user, secret)
		// </mark_2>
	}); err != nil {
		return false, err
	}

	// <mark_2>
	// <mark_3>
	clickSecret, err := restate.Promise[string](ctx, "link-clicked").Result()
	// </mark_3>
	// </mark_2>
	if err != nil {
		return false, err
	}

	return clickSecret == secret, nil
}

func (SignupWorkflow) Click(ctx restate.WorkflowSharedContext, secret string) error {
	// <mark_3>
	return restate.Promise[string](ctx, "link-clicked").Resolve(secret)
	// </mark_3>
}

func main() {
	if err := server.NewRestate().
		Bind(restate.Reflect(SignupWorkflow{})).
		Start(context.Background(), ":9080"); err != nil {
		log.Fatal(err)
	}
}

// <end_here>

File: code_snippets/rust/src/concepts/workflows.rs
use crate::concepts::stubs::{create_user_entry, send_email_with_link, User};
use restate_sdk::prelude::*;

// <start_here>
#[restate_sdk::workflow]
trait SignupWorkflow {
    async fn run(user: Json<User>) -> Result<bool, HandlerError>;
    #[shared]
    async fn click(secret: String) -> Result<(), HandlerError>;
}

struct SignupWorkflowImpl;

impl SignupWorkflow for SignupWorkflowImpl {
    // <mark_1>
    async fn run(
        &self,
        mut ctx: WorkflowContext<'_>,
        Json(user): Json<User>,
    ) -> Result<bool, HandlerError> {
        // workflow ID = user ID; workflow runs once per user
        let user_id = ctx.key();

        // <mark_2>
        ctx.run(|| create_user_entry(&user)).await?;
        // </mark_2>

        // <mark_2>
        let secret = ctx.rand_uuid().to_string();
        ctx.run(|| send_email_with_link(user_id, &user, &secret))
            .await?;
        // </mark_2>

        // <mark_3>
        let click_secret = ctx.promise::<String>("email-link").await?;
        // </mark_3>
        Ok(click_secret == secret)
    }
    // </mark_1>

    // <mark_3>
    async fn click(
        &self,
        ctx: SharedWorkflowContext<'_>,
        secret: String,
    ) -> Result<(), HandlerError> {
        ctx.resolve_promise::<String>("email-link", secret);
        Ok(())
    }
    // </mark_3>
}

#[tokio::main]
async fn main() {
    HttpServer::new(Endpoint::builder().bind(SignupWorkflowImpl.serve()).build())
        .listen_and_serve("0.0.0.0:9080".parse().unwrap())
        .await;
}
// <end_here>
